pkg/restore/actions/csi/volumesnapshotcontent_action.go

```diff
@@ -51,34 +57,77 @@ func (p *volumeSnapshotContentRestoreItemAction) AppliesTo() (
 func (p *volumeSnapshotContentRestoreItemAction) Execute(
 	input *velero.RestoreItemActionExecuteInput,
 ) (*velero.RestoreItemActionExecuteOutput, error) {
-	p.log.Info("Starting VolumeSnapshotContentRestoreItemAction")
-	var snapCont snapshotv1api.VolumeSnapshotContent
 	if boolptr.IsSetToFalse(input.Restore.Spec.RestorePVs) {
 		p.log.Infof("Restore did not request for PVs to be restored %s/%s",
 			input.Restore.Namespace, input.Restore.Name)
 		return &velero.RestoreItemActionExecuteOutput{SkipRestore: true}, nil
 	}
+
+	p.log.Info("Starting VolumeSnapshotContentRestoreItemAction")
+
+	var vsc snapshotv1api.VolumeSnapshotContent
 	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(
-		input.Item.UnstructuredContent(), &snapCont); err != nil {
+		input.Item.UnstructuredContent(), &vsc); err != nil {
 		return &velero.RestoreItemActionExecuteOutput{},
 			errors.Wrapf(err, "failed to convert input.Item from unstructured")
 	}
 
+	var vscFromBackup snapshotv1api.VolumeSnapshotContent
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(
+		input.ItemFromBackup.UnstructuredContent(), &vscFromBackup); err != nil {
+		return &velero.RestoreItemActionExecuteOutput{},
+			errors.Errorf(err.Error(), "failed to convert input.ItemFromBackup from unstructured")
+	}
+
+	// If cross-namespace restore is configured, change the namespace
+	// for VolumeSnapshot object to be restored
+	newNamespace, ok := input.Restore.Spec.NamespaceMapping[vsc.Spec.VolumeSnapshotRef.Namespace]
+	if ok {
+		// Update the referenced VS namespace to the mapping one.
+		vsc.Spec.VolumeSnapshotRef.Namespace = newNamespace
+	}
+
+	// Reset VSC name to align with VS.
+	vsc.Name = util.GenerateSha256FromRestoreUIDAndVsName(
```

Comment:
Should we add a label to the restored VS & VSC object with the original name as well? That might help for debugging.

Reply:
I agree this is useful from the code perspective, but is there any scenario needing that mapping relation?
IMO, the VS and the VSC are not useful after the PVC creation. 
The original proposal was to delete the restored VS and VSC after the restore completes.
As a result, the mapping may not be that useful to me, but if there is any concrete scenario that needs that, I'm open to adding it.

Reply:
No @blackpiglet I don't think of any scenario that needs this. I only asked from code/debug perspective.